src/v/kafka/server/quota_manager.cc

```diff
@@ -53,44 +65,86 @@ ss::future<> quota_manager::stop() {
     // shares a default quota. the anonymous group is keyed on empty string.
-    auto qid = quota_id ? *quota_id : "";
-
-    // find or create the throughput tracker for this client
-    //
-    // c++20: heterogeneous lookup for unordered_map can avoid creation of
-    // an sstring here but std::unordered_map::find isn't using an
-    // equal_to<> overload. this is a general issue we'll be looking at. for
-    // now, these client-name strings are small. This will be solved in
-    // c++20 via Hash::transparent_key_equal.
-    auto [it, inserted] = _client_quotas.try_emplace(
-      ss::sstring(qid),
-      client_quota{
-        now,
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        {static_cast<size_t>(_default_num_windows()), _default_window_width()},
-        /// pm_rate is only non-nullopt on the qm home shard
-        (ss::this_shard_id() == quota_manager_shard)
-          ? std::optional<token_bucket_rate_tracker>(
-            {*_target_partition_mutation_quota(),
-             static_cast<uint32_t>(_default_num_windows()),
-             _default_window_width()})
-          : std::optional<token_bucket_rate_tracker>()});
-
-    // bump to prevent gc
-    if (!inserted) {
-        it->second.last_seen = now;
+    auto qid = quota_id.value_or("");
+
+    vassert(_client_quotas, "_client_quotas should have been initialized");
+
+    auto it = _client_quotas->find(qid);
+    if (it == _client_quotas->end()) {
+        co_await container().invoke_on(
+          _client_quotas.shard_id(),
+          [qid, now](quota_manager& me) { return me.add_quota_id(qid, now); });
+
+        it = _client_quotas->find(qid);
+        if (it == _client_quotas->end()) {
+            // The newly inserted quota map entry should always be available
+            // here because the update to the map is guarded by a mutex, so
+            // there is no chance of losing updates. There is a low chance
+            // that we insert into the map, then there is an unusually long
+            // pause on the handler core, gc() gets scheduled and cleans up
+            // the inserted quota, and by the time the handler gets here the
+            // inserted entry is gone. In that case, we give up and don't
+            // throttle.
+            vlog(klog.debug, "Failed to find quota id after insert...");
+            co_return clock::duration::zero();
+        }
+    } else {
+        // bump to prevent gc
+        it->second->last_seen = now;
     }
 
-    co_return it;
+    co_return cb(*it->second);
+}
+
+ss::future<>
+quota_manager::add_quota_id(std::string_view qid, clock::time_point now) {
+    vassert(
+      ss::this_shard_id() == _client_quotas.shard_id(),
+      "add_quota_id should only be called on the owner shard");
+
+    auto update_func = [this, qid = ss::sstring{qid}, now](
+                         client_quotas_map_t new_map) -> client_quotas_map_t {
```

Comment:
> Oh I guess that's the signature expected. Anyways leaving this here for posterity

Yeah. Either a T&& or T const&, taking a sink param/copy  at call-site to modify in-plaice.

Reply:
It took a copy of the map because in a previous iteration of this PR `update` took an async function. I can revert to `T const&` though now that `update` just takes a normal (non-async) function to simplify.